torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+              b = b->owningNode()->owningBlock();
+            }
+            v_itr = remap[inp].v;
+          }
+
+          while (block_info.at(b) > ancestor_level) {
+            b->registerOutput(v_itr);
+            Value* remapped = b->owningNode()->addOutput();
+            v_itr = remapped;
+            b = b->owningNode()->owningBlock();
+          }
+          remap[inp] = {v_itr, ancestor_level};
+        }
+        TORCH_INTERNAL_ASSERT(remap.count(inp));
+        n->replaceInput(i, remap[inp].v);
+      }
+    }
+
+    if (n->kind() == prim::FakeScopeBlock) {
+      makeDefsDominateUses(n->blocks()[0], remap, block_info);
+    }
+  };
+
+  process_node(b->param_node());
+  for (Node* n : b->nodes()) {
+    process_node(n);
+  }
+  process_node(b->return_node());
+}
+
+void convertReturnsToTuples(Block* b) {
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      TORCH_INTERNAL_ASSERT(n->blocks().size() == 1);
+      convertReturnsToTuples(n->blocks()[0]);
+
+      Graph* g = b->owningGraph();
+      Block* sub_block = n->blocks()[0];
+      if (sub_block->outputs().size() > 1) {
+        {
+          // Make block returns go through a Tuple
+          WithInsertPoint guard(sub_block->return_node());
+          Node* return_tup =
+              g->insertNode(g->createTuple(sub_block->outputs()));
+          while (sub_block->outputs().size()) {
+            sub_block->eraseOutput(0);
+          }
+          sub_block->registerOutput(return_tup->output());
+        }
+
+        // Make node outputs a single tuple;
+        std::vector<TypePtr> types;
+        for (size_t i = 0; i < n->outputs().size(); ++i) {
+          types.push_back(n->output(0)->type());
+        }
+        Value* tup_output = n->addOutput()->setType(TupleType::create(types));
+        Node* tup_unpack = g->createTupleUnpack(tup_output)->insertAfter(n);
+        for (size_t i = 0; i < tup_unpack->outputs().size(); ++i) {
+          n->output(0)->replaceAllUsesWith(tup_unpack->output(i));
+          n->eraseOutput(0);
+        }
+      } else if (sub_block->outputs().size() == 0) {
+        WithInsertPoint guard(sub_block->return_node());
+        sub_block->registerOutput(g->insertNode(g->createNone())->output());
+        n->addOutput()->setType(NoneType::get());
+      }
+    }
+  }
+}
+
+void lambdaLiftBlocksAndConvertToGraph(Block* b) {
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      lambdaLiftBlocksAndConvertToGraph(n->blocks()[0]);
+
+      auto graph = std::make_shared<Graph>();
+      std::unordered_map<Value*, Value*> remaps;
+      graph->block()->cloneFrom(n->blocks()[0], [&](Value* v) {
+        if (!remaps.count(v)) {
+          remaps[v] = graph->addInput()->copyMetadata(v);
+          n->addInput(v);
+        }
+        return remaps[v];
+      });
+      LintGraph(graph);
+      n->g_(attr::Subgraph, graph);
+      n->eraseBlock(0);
+    }
+  }
+
+  // Inline base __module call
```

Comment:
I don't understand what is going on here. Seems like something that should be handled separately. Comment explains what but not why